pgrx/src/callconv.rs

```diff
@@ -283,3 +284,170 @@ where
+impl<'fcx> FcInfo<'fcx> {
+    /// Constructor, used to wrap a raw FunctionCallInfo provided by Postgres.
+    ///
+    /// # Safety
+    ///
+    /// This function is unsafe as we cannot ensure the `fcinfo` argument is a valid
+    /// [`pg_sys::FunctionCallInfo`] pointer.  This is your responsibility.
+    pub unsafe fn assume_valid(fcinfo: pg_sys::FunctionCallInfo) -> FcInfo<'fcx> {
+        let _nullptr_check =
+            std::ptr::NonNull::new(fcinfo).expect("fcinfo pointer must be non-null");
+        Self(fcinfo, std::marker::PhantomData)
+    }
+    /// Retrieve the arguments to this function call as a slice of [`pgrx_pg_sys::NullableDatum`]
+    #[inline]
+    pub fn raw_args<'a>(&'a self) -> &'fcx [pgrx_pg_sys::NullableDatum]
+    where
+        'a: 'fcx,
+    {
+        // Null pointer check already performed on immutable pointer
+        // at construction time.
+        unsafe {
+            let arg_len = (*self.0).nargs;
+            let args_ptr: *const pg_sys::NullableDatum = std::ptr::addr_of!((*self.0).args).cast();
+            // A valid FcInfoWrapper constructed from a valid FuntionCallInfo should always have
+            // at least nargs elements of NullableDatum.
+            std::slice::from_raw_parts(args_ptr, arg_len as _)
+        }
+    }
+
+    /// Modifies the contained `fcinfo` struct to flag its return value as null.
+    ///
+    /// Returns a null-pointer Datum for use with the calling function's return value.
+    ///
+    /// Safety: If this flag is set, regardless of what your function actually returns,
+    /// Postgress will presume it is null and discard it. This means that, if you call this
+    /// method and then return a value anyway, *your extension will leak memory.* Please
+    /// ensure this method is only called for functions which, very definitely, returns null.
+    ///
+    /// # Examples
+    ///
+    /// ```rust,no_run
+    /// use pgrx::callconv::FcInfo;
+    /// use pgrx::prelude::*;
+    /// fn foo<'a>(mut fcinfo: FcInfo<'a>) -> pg_sys::Datum {
+    ///     return unsafe { fcinfo.set_return_null() };
+    /// }
+    /// ```
+    #[inline]
+    pub unsafe fn set_return_null(&mut self) -> pg_sys::Datum {
+        let fcinfo = unsafe { self.0.as_mut() }.unwrap();
+        fcinfo.isnull = true;
+        pg_sys::Datum::from(0)
+    }
+
+    /// Get the collation the function call should use
+    #[inline]
+    pub fn get_collation(&self) -> pg_sys::Oid {
+        // SAFETY: see FcInfo::assume_valid
+        let fcinfo = unsafe { self.0.as_mut() }.unwrap();
+        fcinfo.fncollation
+    }
+
+    /// Retrieve the type (as an Oid) of argument number `num`.
+    /// In other words, the type of `self.raw_args()[num]`
+    #[inline]
+    pub fn get_arg_type(&self, num: usize) -> pg_sys::Oid {
+        // SAFETY: see FcInfo::assume_valid
+        unsafe {
+            pg_sys::get_fn_expr_argtype(self.0.as_ref().unwrap().flinfo, num as std::os::raw::c_int)
+        }
+    }
```

Comment:
Should have an Option signature for the out-of-bound case.

Reply:
ideally there'd be an args iterator that you can use to optionally call `.get_arg_type` for that index. can be deferred to a followup though.

Reply:
This should be resolved, except the iterator thing which I will make into another issue.